File: packages/python/tests/test_regex_parity.py
"""The Python package against its own behaviour, and against conformance/cases.tsv.

The conformance cases are the answers the JavaScript package gave; the Java tests read the same
file. unittest only, so the package needs nothing installed to be tested.

    python3 -m unittest discover -s tests -t .
"""

import re
import unittest
from pathlib import Path

import regex_parity as rp

CONFORMANCE = Path(__file__).resolve().parents[3] / "conformance"


def cp(*code_points: int) -> str:
    return "".join(chr(c) for c in code_points)


class Behaviour(unittest.TestCase):
    def test_a_long_s_does_not_get_past_a_rule(self):
        text = "That is definitely harmle" + cp(0x017F) + "s."
        self.assertTrue(rp.matches(r"\bharmless\b", text))

    def test_find_all_reports_ranges_of_the_original_text(self):
        text = "a mela" + cp(0x200B) + "noma, order " + cp(0xFF11, 0xFF12, 0xFF13)
        found = rp.find_all(r"\bmelanoma\b", text)
        self.assertEqual(found, [rp.Match(2, 11, "mela" + cp(0x200B) + "noma")])
        self.assertEqual(rp.find_all(r"\d{3}", text)[0].text, cp(0xFF11, 0xFF12, 0xFF13))

    def test_case_sensitivity_is_an_option_and_ascii_only_either_way(self):
        self.assertTrue(rp.matches(r"\bTODO\b", "todo"))
        self.assertFalse(rp.matches(r"\bTODO\b", "todo", case_sensitive=True))
        self.assertFalse(rp.matches(r"\bthis\b", "th" + cp(0x0131) + "s"), "a dotless i is not an i")

    def test_an_accented_letter_is_not_a_word_character(self):
        self.assertTrue(rp.matches(r"\bcaf\b", "caf" + cp(0x00E9)))
        self.assertFalse(rp.matches(r"^\w+\s", "caf" + cp(0x00E9) + " ok"))

    def test_unless_only_counts_inside_the_same_sentence(self):
        rule = rp.Rule("date", (r"\bready\s+by\s+friday\b",), (r"\bif\b",))
        self.assertEqual(rp.evaluate(rule, "Ready by Friday if it passes."), [])
        self.assertEqual(len(rp.evaluate(rule, "Ready by Friday. If you ask me, late.")), 1)

    def test_a_range_found_by_two_patterns_counts_once(self):
        rule = rp.Rule("twice", (r"\bharmless\b", "harmless"))
        self.assertEqual(len(rp.evaluate(rule, "harmless")), 1)

    def test_folding_keeps_a_map_back_to_the_original_text(self):
        folded = rp.fold("it" + cp(0x2019) + "s" + cp(0x000D, 0x2028) + "ok " + cp(0x2014) + " fine")
        self.assertEqual(folded.text, "it's\n\nok - fine")
        self.assertEqual(len(folded.starts), len(folded.text))

    def test_the_sentence_function_includes_its_terminator(self):
        text = "One. Two here! Three"
        self.assertEqual(rp.sentence_range(text, 6), (5, 14))
        self.assertEqual(rp.sentence_range(text, 16), (15, 20))

    def test_patterns_engines_read_differently_are_refused(self):
        refused = ["(?=a)", "(?<!a)b", r"(a)\1", "(?P<name>a)", "(?i)a", "a$", r"\p{L}", "a++", "a{,3}", "x{",
                   "[[a]]", "[a&&b]", "[]a]", r"\x41", chr(92) + "u0041", r"\Aa", r"\v", "caf" + cp(0x00E9), r"[\b]"]
        for source in refused:
            self.assertTrue(rp.check_pattern(source), source)
        accepted = [r"\bcolou?r\b", "a{2,3}", "a{2}", r"[a-z\]]", r"\(\?=", "(?:ab)+?", r"\d+\.\d*", "[^.!?]{0,30}", r"\$5"]
        for source in accepted:
            self.assertEqual(rp.check_pattern(source), [], source)
        with self.assertRaises(rp.PatternError):
            rp.compile_pattern("(?=a)")
        with self.assertRaises(rp.PatternError):
            rp.compile_pattern("(a")

    def test_every_variant_folds_back_to_the_original(self):
        original = "It's \"fine\" - order #123456 kept"
        found = rp.variants(original)
        self.assertEqual([v.name for v in found], ["long s", "Kelvin sign", "full-width", "zero-width spaces",
                                                    "no-break spaces", "en dashes", "smart quotes"])
        for variant in found:
            # The Kelvin sign folds to a capital K, so compare without case.
            self.assertEqual(rp.fold(variant.text).text.lower(), rp.fold(original).text.lower(), variant.name)


_ESCAPE = re.compile(re.escape(chr(92)) + "u([0-9a-f]{4})")


def _unescape(text: str) -> str:
    return _ESCAPE.sub(lambda m: chr(int(m.group(1), 16)), text)


def _rows(name: str) -> list[list[str]]:
    lines = (CONFORMANCE / name).read_text(encoding="utf-8").splitlines()
    return [line.split("\t") for line in lines if line and not line.startswith("#")]


class Conformance(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        specs: dict[str, dict] = {}
        for rule_id, case, role, pattern in _rows("rules.tsv"):
            spec = specs.setdefault(rule_id, {"id": rule_id, "case": case, "patterns": [], "unless": []})
            spec["patterns" if role == "pattern" else "unless"].append(_unescape(pattern))
        cls.rules = {rule_id: rp.Rule.from_dict(spec) for rule_id, spec in specs.items()}
        cls.cases = [
            {"rule": r, "expect": e, "variant": v, "text": _unescape(t), "findings": f}
            for r, e, v, t, f in _rows("cases.tsv")
        ]

    def test_every_case_gets_the_recorded_answer_and_findings(self):
        self.assertGreater(len(self.cases), 100)
        for case in self.cases:
            found = rp.evaluate(self.rules[case["rule"]], case["text"])
            label = f"{case['rule']} {case['expect']} ({case['variant']})"
            self.assertEqual(bool(found), case["expect"] == "match", label)
            self.assertEqual(",".join(f"{f.start}:{f.end}" for f in found) or "-", case["findings"], label)

    def test_variants_are_generated_exactly_as_recorded(self):
        groups = []
        for case in self.cases:
            if case["variant"] == "as written":
                groups.append((case, []))
            else:
                groups[-1][1].append((case["variant"], case["text"]))
        for base, expected in groups:
            rule = self.rules[base["rule"]]
            got = [(v.name, v.text) for v in rp.variants(base["text"], rule.case_sensitive)]
            self.assertEqual(got, expected, base["text"])


if __name__ == "__main__":
    unittest.main()

File: packages/ruby/test/test_regex_parity.rb
# frozen_string_literal: true

require "minitest/autorun"
require "regex_parity"

class RegexParityTest < Minitest::Test
  RP = RegexParity
  CONFORMANCE = File.expand_path("../../../conformance", __dir__)

  def cp(*code_points)
    code_points.pack("U*")
  end

  def test_a_long_s_does_not_get_past_a_rule
    assert RP.matches?('\bharmless\b', "That is definitely harmle#{cp(0x017F)}s.")
  end

  def test_find_all_reports_ranges_of_the_original_text
    text = "a mela#{cp(0x200B)}noma, order #{cp(0xFF11, 0xFF12, 0xFF13)}"
    assert_equal [RP::Match.new(2, 11, "mela#{cp(0x200B)}noma")], RP.find_all('\bmelanoma\b', text)
    assert_equal cp(0xFF11, 0xFF12, 0xFF13), RP.find_all('\d{3}', text).first.text
  end

  def test_case_sensitivity_is_an_option_and_ascii_only_either_way
    assert RP.matches?('\bTODO\b', "todo")
    refute RP.matches?('\bTODO\b', "todo", case_sensitive: true)
    refute RP.matches?('\bthis\b', "th#{cp(0x0131)}s"), "a dotless i is not an i"
  end

  def test_a_sharp_s_is_not_two_s
    refute RP.matches?('\bstrasse\b', "stra#{cp(0x00DF)}e"), "Ruby's own /i would match this"
    assert RP.matches?('[a-f]+\d', "ABC1")
  end

  def test_an_accented_letter_is_not_a_word_character
    assert RP.matches?('\bcaf\b', "caf#{cp(0x00E9)}")
    refute RP.matches?('^\w+\s', "caf#{cp(0x00E9)} ok")
    refute RP.matches?('^ok', "caf#{cp(0x00E9)}\nok"), "^ is the start of the text, not of a line"
  end

  def test_unless_only_counts_inside_the_same_sentence
    rule = RP::Rule.new("date", ['\bready\s+by\s+friday\b'], ['\bif\b'])
    assert_empty RP.evaluate(rule, "Ready by Friday if it passes.")
    assert_equal 1, RP.evaluate(rule, "Ready by Friday. If you ask me, late.").length
  end

  def test_a_range_found_by_two_patterns_counts_once
    assert_equal 1, RP.evaluate(RP::Rule.new("twice", ['\bharmless\b', "harmless"]), "harmless").length
  end

  def test_folding_keeps_a_map_back_to_the_original_text
    text = "it#{cp(0x2019)}s#{cp(0x000D, 0x2028)}ok #{cp(0x2014)} fine#{cp(0x000B)}!"
    folded = RP.fold(text)
    assert_equal "it's\n\nok - fine !", folded.text
    start, finish = folded.original_range(2, 3)
    assert_equal cp(0x2019), text[start...finish]
  end

  def test_the_sentence_function_includes_its_terminator
    assert_equal [5, 14], RP.sentence_range("One. Two here! Three", 6)
    assert_equal [15, 20], RP.sentence_range("One. Two here! Three", 16)
  end

  def test_patterns_engines_read_differently_are_refused
    backslash = 92.chr
    refused = ["(?=a)", "(?<!a)b", '(a)\1', "(?<name>a)", "(?i)a", "a$", '\p{L}', "a++", "a{,3}", "x{", "[[a]]", "[a&&b]",
               "[a--b]", "[a~~b]", "[]a]", '\x41', "#{backslash}u0041", '\Aa', '\v', "caf#{cp(0x00E9)}", '[\b]', "a{1001}", "a{2,5000}"]
    refused.each { |source| refute_empty RP.check_pattern(source), source }
    accepted = ['\bcolou?r\b', "a{2,3}", "a{2}", '[a-z\]]', '\(\?=', "(?:ab)+?", '\d+\.\d*', "[^.!?]{0,30}", '\$5', "a{1000}"]
    accepted.each do |source|
      assert_empty RP.check_pattern(source), source
      RP.compile(source)
    end
    assert_raises(RP::PatternError) { RP.compile("(?=a)") }
    assert_raises(RP::PatternError) { RP.compile("(a") }
  end

  def test_every_variant_folds_back_to_the_original
    original = %(It's "fine" - order #123456 kept)
    variants = RP.variants(original)
    assert_equal ["long s", "Kelvin sign", "full-width", "zero-width spaces", "no-break spaces", "en dashes", "smart quotes"], variants.map(&:name)
    # The Kelvin sign folds to a capital K, so compare without case.
    variants.each { |v| assert_equal RP.fold(original).text.downcase, RP.fold(v.text).text.downcase, v.name }
  end

  # --- conformance: the same answers as conformance/*.tsv, which every other package also reads ---

  def unescape(text)
    text.gsub(/#{Regexp.escape(92.chr)}u([0-9a-f]{4})/) { [Regexp.last_match(1).to_i(16)].pack("U") }
  end

  def rows(name)
    File.readlines(File.join(CONFORMANCE, name), chomp: true, encoding: "UTF-8")
        .reject { |line| line.empty? || line.start_with?("#") }
        .map { |line| line.split("\t", -1) }
  end

  def conformance_rules
    rules = {}
    rows("rules.tsv").each do |id, kase, role, pattern|
      rule = rules[id] ||= RP::Rule.new(id, [], [], kase == "sensitive")
      (role == "pattern" ? rule.patterns : rule.unless) << unescape(pattern)
    end
    rules
  end

  def test_conformance_every_case_gets_the_recorded_answer_and_findings
    rules = conformance_rules
    cases = rows("cases.tsv")
    assert_operator cases.length, :>, 100
    failures = cases.filter_map do |rule, expect, variant, text, findings|
      found = RP.evaluate(rules[rule], unescape(text))
      spans = found.empty? ? "-" : found.map { |f| "#{f.start}:#{f.end}" }.join(",")
      "#{rule} #{expect} (#{variant}): got #{spans}, want #{findings}" if (!found.empty?) != (expect == "match") || spans != findings
    end
    assert_empty failures, failures.join("\n")
  end

  def test_conformance_variants_are_generated_exactly_as_recorded
    rules = conformance_rules
    groups = []
    rows("cases.tsv").each do |rule, _expect, variant, text, _findings|
      if variant == "as written"
        groups << [rule, unescape(text), []]
      else
        groups.last[2] << [variant, unescape(text)]
      end
    end
    groups.each do |rule, base, expected|
      got = RP.variants(base, case_sensitive: rules[rule].case_sensitive).map { |v| [v.name, v.text] }
      assert_equal expected, got, base
    end
  end

  def test_conformance_every_language_refuses_and_accepts_the_same_patterns
    patterns = rows("patterns.tsv")
    assert_operator patterns.length, :>, 10
    patterns.each do |expect, source|
      assert_equal expect == "refused", !RP.check_pattern(unescape(source)).empty?, source
    end
  end
end
